app/notification_push.go

```diff
@@ -270,20 +314,45 @@ func (a *App) StopPushNotificationsHubWorkers() {
 	}
 }
 
-func (a *App) sendToPushProxy(msg model.PushNotification, session *model.Session) {
+func (a *App) sendToPushProxy(msg model.PushNotification, session *model.Session) error {
+	pushNotificationFailed := func(notification *model.PushNotification, err error) {
+		if pluginsEnvironment := a.GetPluginsEnvironment(); pluginsEnvironment != nil {
+			pluginContext := a.PluginContext()
+			pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+				hooks.PushNotificationHasFailed(pluginContext, notification, err)
+				return true
+			}, plugin.PushNotificationHasFailedId)
+		}
+		mlog.Debug("Failed to send Push notification", mlog.String("notificationId", notification.Id), mlog.String("ackId", notification.AckId), mlog.String("deviceId", notification.DeviceId))
+	}
+
 	msg.ServerId = a.DiagnosticId()
 
-	request, err := http.NewRequest("POST", strings.TrimRight(*a.Config().EmailSettings.PushNotificationServer, "/")+model.API_URL_SUFFIX_V1+"/send_push", strings.NewReader(msg.ToJson()))
+	var updatedMsg *model.PushNotification
+
+	if pluginsEnvironment := a.GetPluginsEnvironment(); pluginsEnvironment != nil {
+		pluginContext := a.PluginContext()
+		pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+			updatedMsg = hooks.PushNotificationWillBeSent(pluginContext, &msg)
+			return true
+		}, plugin.PushNotificationWillBeSentId)
+	}
+	mlog.Debug("Push notification will be sent", mlog.String("notificationId", updatedMsg.Id), mlog.String("ackId", updatedMsg.AckId), mlog.String("deviceId", updatedMsg.DeviceId))
+
+	if updatedMsg == nil {
+		return nil
```

Comment:
Maybe add a debug message here? This would help plugins devs.

Reply:
Yes, makes sense. Something like "Push notification inhibited from plugin".